network/parallelism.go

```diff
@@ -185,11 +187,26 @@ func (p *parallelism) desiredLoop() {
 	}
 	// If we are drifting too much then ramp up the number of loops.
 	if float64(p.timestampDriftSeconds) > p.cfg.AllowedDrift.Seconds() {
-		// Need to keep the value between min and max.
-		if p.currentDesired+1 <= p.cfg.MaxConnections {
-			level.Debug(p.l).Log("msg", "increasing desired due to timestamp drift", "desired", p.currentDesired+1, "drift", p.timestampDriftSeconds)
-			p.calculateDesiredParallelism(p.currentDesired + 1)
+		// Check if drift has been increasing by looking at previous desired values
+		var increaseFactor uint = 1
+
+		if len(p.previous) >= 2 {
+			// Look at previous values in chronological order to see if we should be more aggressive
+			maxPrevious := slices.MaxFunc(p.previous, func(a, b previousDesired) int {
+				return cmp.Compare(a.desired, b.desired)
+			})
+			// If the most recent previous value was the maximum, that means drift is consistently increasing
```

Comment:
I have mixed feelings about this. Is there a way to calculate the required parallelism using throughput numbers? For example:
- if we're sending 100k samples/s, but WAL is backing up at a rate of 20k samples/s, then required throughput is 120k samples/s.
- From the above we can estimate that current parallelism needs to increase by 20%. Say it's 20 loops -> we increase it to 24.

I worry the current version may not respond well to some traffic patterns and we will need another update here soon, which may in turn break other traffic pattern and it may become a bit of a moving target.

Reply:
We can its a bit more complicated. Let me see if I can get something that makes sense.